Add SmallRye GraphQL extension

phillip-kruger opened the pull request:
This PR adds Smallrye GraphQL (a MicroProfile GraphQL Implementation) to Quarkus.

This PR will also resolve #2509

jmartisk commented:
As a side note, I have SmallRye Metrics integration with SmallRye GraphQL almost ready too, it should arrive soon after this is merged.

gsmet commented on bom/runtime/pom.xml:
```diff
@@ -2968,7 +2995,7 @@
                 <version>${swagger-ui.version}</version>
             </dependency>
-
+            
```
This one doesn't look necessary :).

gsmet commented on extensions/pom.xml:
```diff
@@ -158,9 +158,12 @@
+
+	<!-- GraphQL -->
         <module>vertx-graphql</module>
+	<module>smallrye-graphql</module>
```
Indentation issues.

gsmet commented on build-parent/pom.xml:
```diff
@@ -68,7 +68,8 @@
         <microprofile-context-propagation.version>1.0.1</microprofile-context-propagation.version>
         <microprofile-jwt-api.version>1.1.1</microprofile-jwt-api.version>
-
+        <microprofile-graphql.version>1.0.2</microprofile-graphql.version>
```
Maybe add the `-api` suffix? Looks like they are all consistent except context-propagation.

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+ * Processor for Smallrye GraphQL.
+ * We scan all annotations and build the model during build.
+ * 
+ * @author Phillip Kruger (phillip.kruger@redhat.com)
```
We don't do `@author` tags in Quarkus and rely on Git history.

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+    private static final String SPI_PATH = "META-INF/services/";
+
+    @Inject
+    private LaunchModeBuildItem launchMode;
```
Better push that as a parameter of your methods. It's better to keep track of dependencies.

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+    SmallRyeGraphQLConfig quarkusConfig;
+
+    @ConfigRoot(name = "smallrye-graphql")
+    static final class SmallRyeGraphQLConfig {
```
I would make that a separate class.

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+        Set<String> metricsImplementations = ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
+                metricsService);
+        serviceProvider.produce(
+                new ServiceProviderBuildItem(MetricsService.class.getName(), metricsImplementations.toArray(new String[0])));
```
Is this related to SmallRye Metrics, should we test the presence of it?

jmartisk replied:
Metrics integration is not supposed to be covered by this PR yet, I am planning to add that later, so I think this is some leftover that can be removed for now

phillip-kruger replied:
No, I added it to load the SPI (Native mode). @jmartisk will this use another service implementation ? I can remove this, then you add it when you do the metrics ?

jmartisk replied:
Well, yeah, this is just one part of the metrics integration - without the rest, the metrics won't work properly anyway. So yeah I'd say we can remove these six lines for now, I'll do metrics separately later and re-add this

phillip-kruger commented:
Documentation PR here: https://github.com/quarkusio/quarkus/pull/8967

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+         * This is because the worker thread on which the handler is called captures the TCCL at creation time
+         * and does not allow updating it.
+         *
+         * In non dev mode, the TCCL is used.
```
This is being worked on by @stuartwdouglas IIUC the issue.

phillip-kruger replied:
Yes, once his changes is in I can remove this and the special QuarkusClassloadingService class

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+        requireBodyHandlerProducer.produce(new RequireBodyHandlerBuildItem());
+
+        // TODO: SSL Support by default ?
+        // extensionSslNativeSupportProducer.produce(new ExtensionSslNativeSupportBuildItem(FeatureBuildItem.SMALLRYE_GRAPHQL));
```
Why would the extension need SSL? It connects to something?

gsmet commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,434 @@
+    private static final String GRAPHQL_UI_FINAL_DESTINATION = "META-INF/graphql-ui-files";
+    private static final String TEMP_DIR_PREFIX = "quarkus-graphql-ui_" + System.nanoTime();
+    private static final List<String> IGNORE_LIST = Arrays.asList(new String[] { "logo.png", "favicon.ico" });
+    private static final String FILE_TO_UPDATE = "render.js";
```
This should definitely be at the top of the class.

gsmet commented on extensions/smallrye-graphql/runtime/pom.xml:
```diff
@@ -0,0 +1,76 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-smallrye-graphql</artifactId>
+    <name>Quarkus - SmallRye GraphQL - Runtime</name>
```
This module needs a description, it's used in code.quarkus.io. Better start the description with a verb "Connect to..." or whatever :).

gsmet commented on extensions/smallrye-graphql/deployment/pom.xml:
```diff
@@ -0,0 +1,83 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jsonb-deployment</artifactId>
+        </dependency>
```
The jsonp deployment is missing here. It should be symmetrical with the runtime pom. As this dependency comes with jsonb, I would simply remove it from the runtime module

gsmet commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    private static final String QUERY = "query";
+    private static final String OK = "OK";
```
Please put the constants at the top of the class.

gsmet commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:
```diff
@@ -0,0 +1,82 @@
+import io.vertx.ext.web.handler.StaticHandler;
+
+/**
+ * Recorder
```
Let's drop useless javadoc comments.

gsmet commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:
```diff
@@ -0,0 +1,82 @@
+public class SmallRyeGraphQLRecorder {
+
+    public void createExecutionService(Schema schema) {
+        Arc.initialize(); // FIXME: why do I need this?
```
This is incorrect. You probably want a `BeanContainerListener`.

phillip-kruger replied:
Thanks :)

gsmet commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java:
```diff
@@ -0,0 +1,57 @@
+ * 
+ * @author Phillip Kruger (phillip.kruger@redhat.com)
+ */
+public class QuarkusClassloadingService implements ClassloadingService {
```
This needs @stuartwdouglas 's attention.

phillip-kruger replied:
Yes, this is the hack until @stuartwdouglas 's fix is in. Same as open-api extension. I'll update both once the fix is in.

phillip-kruger replied:
Can I resolve this ? As I said as soon as there is a fix I will remove the hack here and in open-api ? @gsmet  ?

gsmet commented on tcks/microprofile-graphql/pom.xml:
```diff
@@ -0,0 +1,48 @@
+            <artifactId>microprofile-graphql-tck</artifactId>
+            <version>${microprofile-graphql.version}</version>
+            <scope>provided</scope>
+        </dependency>
```
Please make the indentation consistent.

gsmet reviewed (changes requested):
Cool! I added a bunch of comments, mostly on the form. Could you address them so that we get these out of the way and can focus on the implementation?

Thanks!

jmartisk commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,384 @@
+        }
+    }
+
+    private byte[] streamToByte(InputStream is) throws IOException {
```
Could we use `org.apache.commons.io.IOUtils.toByteArray(InputStream)` instead? It seems to be somewhere among the dependencies and some other extensions use it too

phillip-kruger replied:
Happy to use that, did not want to introduce an new dependency, but if it's there it makes sense. Let me know.

gsmet replied:
Please check before doing that that the dependency is not there by pure luck (i.e. brought by another one that is not really related to what we do and where it could go away)

phillip-kruger replied:
Seems like it's available:
GroupId: commons-io 
ArtifactId: commons-io 
Version: 2.6 
Type: jar Scope: compile 

Via: maven-shared-utils 
Via: maven-embedder 
Via: quarkus-bootstrap-core 
Via: quarkus-core-deployment

Should I go ahead and use that ? @gsmet ?

gsmet replied:
Given it's coming incidentally from Maven, I wouldn't use it.

Better add what you need to `io.quarkus.deployment.util.IoUtil`.

phillip-kruger replied:
No need to add anything, the method was there already. Doing a fix now

jmartisk commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,376 @@
+    private static final List<String> IGNORE_LIST = Arrays.asList(new String[] { "logo.png", "favicon.ico" });
+    private static final String FILE_TO_UPDATE = "render.js";
+
+    SmallRyeGraphQLConfig quarkusConfig;
```
I am getting a compile error now, seems you forgot to add this class after extracting it

phillip-kruger replied:
Sorry. (But it build on my machine :))
I'll commit now with the other Arc fix

phillip-kruger replied:
@jmartisk please check now

jmartisk commented:
No further issues from my side at the moment. Tested it with the latest GraphQL snapshot and it works fine, except the introspection in UI in native mode, but I don't really know how to fix that.

phillip-kruger commented:
Ready for review :)

gsmet commented:
@phillip-kruger is there anything still not working? Not saying it should block this PR. I just would like to have a clear status?

Could you squash all the commits into one?

phillip-kruger commented:
Yes, there is a weird intermittent error after a hot reload. I am looking at it now, with the help of clement.
Then the UI is not working properly in native mode. You can do queries, but introspection is not working. I will look at that too a.s.a.p

phillip-kruger commented:
@gsmet :  Squashed ! :)

stuartwdouglas commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,373 @@
+            // Now add our own logo and favicon
+            ClassLoader classLoader = SmallryeGraphqlProcessor.class.getClassLoader();
+            for (String ownMedia : IGNORE_LIST) {
+                InputStream logo = classLoader.getResourceAsStream(OWN_MEDIA_FOLDER + ownMedia);
```
This should be in a try with resources block.

stuartwdouglas commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:
```diff
@@ -0,0 +1,104 @@
+                                .setStatusMessage(OK)
+                                .end(Buffer.buffer(getResponse));
+                    } else {
+                        response.setStatusCode(204).setStatusMessage("Provide a query parameter").end();
```
I don't think you need to bother with the status messages, they are not supported with HTTP/2 and are generally considered to have been a bad idea. We don't want to encourage their use as they are basically deprecated.

stuartwdouglas commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:
```diff
@@ -0,0 +1,104 @@
+        try (StringReader input = new StringReader(body);
+                final JsonReader jsonReader = Json.createReader(input)) {
+            JsonObject jsonInput = jsonReader.readObject();
+            ExecutionService executionService = CDI.current().select(ExecutionService.class).get();
```
You probably want to cache this in a volatile variable. CDI.current().select() is relatively expensive.

mkouba replied:
In fact, it does not have to be a `volatile` field but a final field initialized in the constructor. Also `Arc.container().instance()` is a bit more effective.

phillip-kruger replied:
Did a lazy load with volatile variable.

stuartwdouglas commented on extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:
```diff
@@ -0,0 +1,73 @@
+public class SmallRyeGraphQLRecorder {
+
+    public void createExecutionService(BeanContainer beanContainer, Class<GraphQLProducer> beanClass, Schema schema) {
+        GraphQLProducer graphQLProducer = beanContainer.instance(beanClass);
```
Why do you pass the bean class in like this, when it is hard coded anyway? Why not just use GraphQLProducer.class?

phillip-kruger commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:
```diff
@@ -0,0 +1,373 @@
+        }
+
+        // Make sure the complex object from the application can work in native mode
+        reflectiveClassProducer
```
Cool, thanks. Did not know about this.

gsmet commented:
I don't see any integration tests or am I missing something? We need some to get native covered by our CI. Also it needs to be added to the github CI workflow.

I don't think the TCK tests cover native, right?

phillip-kruger commented:
@gsmet - I fixed all the review items from @stuartwdouglas and @mkouba , then squashed, but made a mistake (fixed now) - however @boring-cyborg added a bunch of labels after my mistake, that I can not remove. 

Can you please remove the following labels:  

- area/arc 
- area/devtools 
- area/documentation 
- area/maven 
- area/mongodb 
- area/panache 
- area/gradle  

Thanks and apologies.

gsmet commented:
OK, I think the only thing blocking this PR is that we need to have test coverage for native. I haven't seen an answer to my question? Right now, I don't think it's covered.

phillip-kruger commented:
Hi @gsmet - Busy adding this as we speak. Should have it ready later today

phillip-kruger commented:
@gsmet - Integration tests with native added. Please let me know if this is ok. Thanks :)

jmartisk commented on integration-tests/smallrye-graphql/nbactions.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file shouldn't be here

phillip-kruger replied:
Sorry about this - removed the file and added Netbeans files to gitignore

phillip-kruger committed:
Adding SmallRye GraphQL Extension

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

gsmet committed:
Add GraphQL to the native workflow and reorganize things a bit

Also the Vault native tests weren't running at all. I hope they are not
failing...

gsmet commented:
@phillip-kruger I added a commit to add the native tests to the CI workflow. It's not something automatic. Let's see what CI has to say and then merge.

gsmet commented:
I fixed a conflict in the CI set up and force pushed. Hopefully I will be able to merge it right after the CI run.

gsmet commented:
@phillip-kruger @jmartisk merged, thanks! I let you continue with follow-up PRs?

gsmet commented:
Also, we need a proper guide and quickstart. I know there is some work in progress but we need to finish that work before the end of the month.

phillip-kruger commented:
Thanks @gsmet . Yes we are working on guide and quick start with @cemnura . Will get it done by the end of the month. Thanks!
